cms/server/admin/handlers/task.py

```diff
@@ -502,3 +508,43 @@ def delete(self, task_id):
 
         # Maybe they'll want to do this again (for another task)
         self.write("../../tasks")
+
+
+class DownloadSubmissionsHandler(FileHandler):
+    """Download submissions for a dataset in a zip file.
+    """
+    @require_permission(BaseHandler.AUTHENTICATED)
+    def get(self, task_id):
+        task = self.safe_get_item(Task, task_id)
+
+        zip_filename = "%s.zip" % task.name
+        source_template = "%s_%s%s"
+
+        tempdir = tempfile.mkdtemp(dir=config.temp_dir)
+        zip_path = os.path.join(tempdir, zip_filename)
+
+        users = {}
+
+        with zipfile.ZipFile(zip_path, "w") as zip_file:
+            for submission in task.submissions:
+                user = submission.participation.user.username
+                if user not in users:
+                    users[user] = 1
+                else:
+                    users[user] += 1
+                for _, sub_file in submission.files.items():
+                    source_path = self.application.service.file_cacher.\
+                        get_file(sub_file.digest).name
+                    zip_file.write(
+                        source_path,
+                        source_template % (
+                            submission.participation.user.username,
+                            users[user],
+                            get_language(submission.language).source_extension
```

Comment:
I didn't manage to test this yet, but to the best of my knowledge submission.language is None for certain submissions such as submissions to an OutputOnly problem which will make this line to raise a ValueError.